### DISCUSSION TITLE
Does daft use all cores by default?

### DISCUSSION THREAD

**Discussion by pawarbi** [Eventual-Inc/Daft#2611] [Q&A]
I am running some tests in Fabric with and without a ray cluster. Am I correct that default daft uses a single core (just like pandas) and daft + ray uses all available cores (by default). Like spark, do I need to partition the data to utilize the cores effectively ? (I did with and without partitioning and saw a big improvement).

**Comment by jaychia** [comment]
> I am running some tests in Fabric with and without a ray cluster. Am I correct that default daft uses a single core (just like pandas) and daft + ray uses all available cores (by default). Like spark, do I need to partition the data to utilize the cores effectively ? (I did with and without partitioning and saw a big improvement). 

Daft by default (without Ray) already uses all cores. Today, it does require partitioning in order to do so more effectively, but Daft will already make efficient use of CPUs for tasks like parallel data loading. We are working on a new execution model that make this even more efficient without partitioning (will be released in about a month or so). 

Daft + Ray also uses all cores, only if your data is partitioned. The main difference is that this mode also allows for out-of-core processing, and distributed processing. Using Ray introduces some overhead, but the overhead is worth it if the data is large. 

In summary, because we optimized for distributed computing, today Daft relies on partitioning to efficiently utilize resources on both the local and Ray runners. We have a new execution engine in the works which will make this easier without partitioning.

**Reply by pawarbi** [reply]
Thanks @jaychia . Makes sense. I did partition in my tests but didn't see much improvement, I will test again. If you have any suggestions, please let me know. 

https://fabric.guru/quick-test-daft-with-ray-in-fabric

**Reply by jaychia** [reply]
Yes indeed, in this case your data is likely already big-enough that Daft will automatically partition it when reading it in. Thus you may not be seeing a large gain in performance because your table was already correctly partitioned in the first place :)

**Reply by jaychia** [reply]
One way you can check is to run `df.explain(True)` to take a look at the physical plan, and inspect the number of partitions

**Reply by caican00** [reply]
> We have a new execution engine in the works which will make this easier without partitioning

Hi @jaychia, may I ask what this new execution engine is and what its advantages are?

**Reply by plotor** [reply]
> > We have a new execution engine in the works which will make this easier without partitioning
> 
> Hi @jaychia, may I ask what this new execution engine is and what its advantages are?

@caican00 The new local engine is called __Swordfish__, and the distributed engine is called __Flotilla__. I see you work in China, so you can read the following two articles published on WeChat to learn more, or follow the Daft Chinese community WeChat account. Also, welcome to join the community WeChat group for discussion.

- https://mp.weixin.qq.com/s/v3tk0l8KBk22LqyuP82tCQ
- https://mp.weixin.qq.com/s/KoG0UWMKa029Hx-Jn9165A
